addon/mixins/ajax-request.ts

```diff
@@ -333,81 +359,48 @@ export default Mixin.create({
 
   /**
    * calls `request()` but forces `options.type` to `POST`
-   *
-   * @method post
-   * @public
-   * @param {string} url The url to make a request to
-   * @param {Object} options The options for the request
-   * @return {Promise} The result of the request
    */
-  post(url, options) {
+  post(url: string, options?: AJAXOptions): AJAXPromise<Response> {
```

Comment:
See my typings where I have generic variants for all the verbs, so for example;

```
        post(url: string, options?: JQuery.AjaxSettings): RSVP.Promise<any>;
        post<T>(url: string, options?: JQuery.AjaxSettings): RSVP.Promise<T>;
```

This allows for clean code on the consumer's side, as the return type is already defined.

Reply:
Makes sense to me, but could you give an example of what that would look like in use? I thought about going that direction with the type definitions, but decided against it because the return value is always going to be some kind of JSON object -- does this just allow the user to say something like this?

```typescript
interface User {
  name: string;
}

const user = await get<User>(this, 'ajax').get('/users/1');
```

and allow their application to know that `user` is a `string`, and give the right Intellisense?

My reservation here is that we have no way to know that `user` is _actually_ matching the shape of the `User` interface; we could try to tell TypeScript that that's the case, but we have no way to verify it. I think it would actually be safer to keep the `AJAXPromise<Response>` return type where `Response` is `any` because it's closer to what we can actually guarantee is true about the system.

Reply:
On the other hand, we _could_ take a "user at your own risk" attitude toward the generic type with something like this:

```typescript
post<T = Response>(url: string, options?: whatever): AjaxPromise<T>
```

since that would basically tell people that, if left unspecified, the return value is going to resolve to `any`. However, if they _want_ to tell the compiler what the type should be treated as, they can -- at their own risk, because they may actually be wrong and cause a runtime error that TypeScript can't help them with.

I think I like this idea best, since it is kind of the "best of both worlds" -- if a user wants to opt into this feature, they can, or else fall back to the original behavior I described, which is arguably safer but might be less helpful.

Reply:
Alright, that's what I did! Thanks for suggesting the generic.

I like the option of having it, if the user is confident in the structure of their JSON response and wants to get the richer experience, but by default it will be of type `Response` (which is an alias for `any`).

What's nice is that `request` and `raw` can both be parameterized by a type, even though `raw` actually has a return type of `AjaxPromise<RawRequest<T>>`. Adding in the generics was fun!